### DISCUSSION TITLE
[Idea] Auto exclude classes tag with some attributes

### DISCUSSION THREAD

**Discussion by Nayte91** [symfony/symfony#59280] [Q&A] [answered]
Hi there,

I remember that long time ago I was told by the framework not to register as a service the Entity objects from Doctrine; And as I see in services.yaml, it's a thing today.

Then, I use to work with Messenger component and its very Command/Query/Event objects, to work with various DTOs, ... And I assume that's also a good idea to withdraw all those Value Objects from the Service Container. So yeah I can add those folders to services.yaml --> services.app\.exclude and go with it.

With last SF 7.2 feature "AsMessage" attribute, I was wondering if there is a way, for some attribute-stamped classes, if they can be automatically excluded from service list?
- AsMessage,
- ORM\Entity,
- Maybe others

1. 1st advantage of auto exclude some attributes like this is for people who forget to exclude in conf; Saves few bytes in RAM and saves some bugs.
2. 2nd advantage if one tags all his VOs with those attributes, he can get rid of another part of the config files, that is always a nice to have.

I wonder we can automatically add the Symfony\Component\DependencyInjection\Attribute\Exclude attribute to AsMessage and ORM\Entity? Or ensure the same processing, on the injector side, for those 2 attributes as for Exclude?

**Comment by Nayte91** [comment]
I'm taking a look to achieve this, but:
- If I want to [add a condition](https://github.com/symfony/dependency-injection/blob/a475747af1a1c98272a5471abc35f3da81197c5d/Loader/FileLoader.php#L154) about the Symfony\Component\Messenger\Attribute\AsMessage::class & Doctrine\ORM\Mapping\Entity attributes::class, then it needs to have dependency to Doctrine URL & Messenger, and we maybe don't want them,
- If I want to add the condition with stringified FQCN like 'Symfony\Component\Messenger\Attribute\AsMessage', it may works but it will need the dependency during tests? SO it will need to add them in require-dev?
- Adding the #[exclude] attribute to Messenger will need to add the dependency, and more over for Doctrine's Entity attribute.

tl;dr: I don't know what is the good practice to handle those dependency issues!

**Reply by GromNaN** [reply]
The class don't need to be defined to call `ReflectionClass::getAttributes($class)` and the `\Symfony\Component\Messenger\Attribute\AsMessage::class` doesn't require the class the be defined either. So don't worry, it doesn't add any dependency.

You can open a PR. I think it's a good idea.

**Comment by nicolas-grekas** [comment, marked as answer]
```php
        $container->registerAttributeForAutoconfiguration(MyExcludingAttribute::class, static function (ChildDefinition $definition): void {
            $definition->addTag('container.excluded');
        });
```
